ksqldb-engine/src/main/java/io/confluent/ksql/engine/RuntimeAssignor.java

```diff
@@ -57,11 +63,28 @@ public RuntimeAssignor createSandbox() {
     return new RuntimeAssignor(this);
   }
 
-  public String getRuntimeAndMaybeAddRuntime(final QueryId queryId,
-                                             final Collection<SourceName> sources,
-                                             final KsqlConfig config) {
-    if (idToRuntime.containsKey(queryId)) {
-      return idToRuntime.get(queryId);
+  synchronized public String getRuntimeAndMaybeAddRuntime(final QueryId queryId,
+                                                          final Collection<SourceName> sources,
+                                                          final KsqlConfig config) {
+    final String runtimeIdForExistingQuery = idToRuntime.get(queryId);
+    if (runtimeIdForExistingQuery != null) {
+      final Set<SourceName> existingSources = runtimesToSources.get(runtimeIdForExistingQuery);
+      final Set<SourceName> overlappingSources =
+          new HashSet<>(existingSources);
+      overlappingSources.retainAll(sources);
+      if (!overlappingSources.isEmpty()) {
+        log.error("Attempt to replace query {} in runtime {} with new input topics {} failed since"
+                      + " the runtime is already consuming from the following input topics: {}\n"
+                      + "Overlapping sources: {}",
+                  queryId, runtimeIdForExistingQuery, sources, existingSources, overlappingSources);
+        throw new IllegalStateException(String.format(
+            "Unable to update query %s to read from stream(s) %s as there is another query in this"
+                + " runtime already consuming from these topic(s). This restriction will be lifted"
+                + " in the near future, but you can work around it for now by dropping the query"
+                + "  and restarting it instead of using CREATE OR REPLACE",
```

Comment:
FWIW, this isn't a workaround - I'm not sure we should suggest it. You can't do this if any downstream query is using the source (which is the only reason to really use COR in first place)